<?xml version="1.0"?>
<robot name="autoBot">

  <material name="black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <material name="red">
    <color rgba="1 0 0 1.0"/>
  </material>
  <material name="green">
    <color rgba="0 1 0 1.0"/>
  </material>
  <material name="blue">
    <color rgba="0 0 1 1.0"/>
  </material>
  <material name="yellow">
    <color rgba="1 1 0 1.0"/>
  </material>
  <material name="white">
    <color rgba="1 1 1 1.0"/>
  </material>

  <!-- base_footprint is a fictitious link(frame) that is on the ground right below base_link origin, navigation stack depends on this frame -->

  <link name="base_footprint">
  </link>

  <joint name="base_link_joint" type="fixed">
    <!-- place the joint at the top of the platform, to make it easy to reference to -->
    <origin xyz="0 0 0.054" rpy="0 0 0" />        
    <parent link="base_footprint"/>
    <child link="base_link" />
  </joint>

  <link name="base_link">
    <visual>
      <!-- this origin is with respect to the joint origin -->
      <origin xyz=" 0 0 -0.002" rpy="0 0 0" />
      <geometry>
        <box size="0.41 0.41 0.004"/>
      </geometry>
      <material name="black"/>
    </visual>
<!-- to reduce processing, we will do all the collision structure from just base link
      we will make it a 12 cm cylinder, starting at the
      baselink joint -->
    <collision>
      <origin xyz="0.0 0.0 0.06" rpy="0 0 0" />
      <geometry>
        <box size="0.41 0.41 0.12"/>
      </geometry>
    </collision>
  </link>

  <joint name="second_level_joint" type="fixed">
    <origin xyz="0 0 0.065" rpy="0 0 0" />        
    <parent link="base_link"/>
    <child link="second_level_link" />
  </joint>


  <link name="second_level_link">
    <visual>
      <origin xyz=" 0 0 -0.002" rpy="0 0 0" />
      <geometry>
        <box size="0.41 0.41 0.004"/>
      </geometry>
      <material name="black"/>
    </visual>

  </link>

  <joint name="third_level_joint" type="fixed">
    <origin xyz="0.02 0 0.108" rpy="0 0 0" />        
    <parent link="second_level_link"/>
    <child link="third_level_link" />
  </joint>


  <link name="third_level_link">
    <visual>
      <origin xyz=" 0 0 -0.002" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.203" length="0.005"/>
      </geometry>
      <material name="black"/>
    </visual>

  </link>

  <joint name="fourth_level_joint" type="fixed">
    <origin xyz="0.025 0 0.26" rpy="0 0 0" />        
    <parent link="third_level_link"/>
    <child link="fourth_level_link" />
  </joint>


  <link name="fourth_level_link">
    <visual>
      <origin xyz=" 0 0 -0.002" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.178" length="0.005"/>
      </geometry>
      <material name="black"/>
    </visual>

  </link>
 <!-- for rotated structures, don't rotate the joint, just rotate the visual representation
      Rotating the joint makes it difficult to figure out things like what axis to use
      and also propates to the link parameters, so it is hard to get things in the right orientation -->
  <joint name="left_wheel_joint" type="continuous">
    <origin xyz="0.11 0.159 0.022" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="left_wheel_link"/>
    <axis xyz="0 1 0"/>
  </joint>


  <link name="left_wheel_link">
    <visual>
      <origin xyz="0 -0.0125 0" rpy="1.5707 0 0" />
      <geometry>
        <cylinder radius="0.076" length = "0.025"/>
      </geometry>
      <material name="red"/>
    </visual>

  </link>

  <joint name="right_wheel_joint" type="continuous">
    <origin xyz="0.11 -0.159 0.022" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="right_wheel_link"/>
    <axis xyz="0 1 0"/>
  </joint>

  <link name="right_wheel_link">
    <visual>
      <origin xyz="0 -0.0125 0" rpy="-1.5707 0 0" />
      <geometry>
        <cylinder radius="0.076" length = "0.025"/>
      </geometry>
      <material name="green"/>
    </visual>
  </link>

  <!-- fixed instead of continuous because there's no transmission -->
  <joint name="rear_wheel_stalk_joint" type="fixed">
    <origin xyz="-0.125 0 -0.081" rpy="0 0 0"/>
    <parent link="second_level_link"/>
    <child link="rear_castor_stalk_link"/>
    <axis xyz="0 0 1"/>
  </joint>

  <link name="rear_castor_stalk_link">
    <visual>
      <origin xyz="0 0 0.0405" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.006" length = "0.081"/>
      </geometry>
      <material name="blue"/>
    </visual>
  </link>

  <!-- fixed instead of continuous because there's no transmission -->
  <joint name="rear_castor_joint" type="fixed">
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <parent link="rear_castor_stalk_link"/>
    <child link="rear_castor_link"/>
    <axis xyz="0 1 0"/>
  </joint>

  <link name="rear_castor_link">
    <visual>
      <origin xyz="0 0 0" rpy="1.5707 0 0" />
      <geometry>
        <cylinder radius="0.038" length = "0.04"/>
      </geometry>
      <material name="white"/>
    </visual>
  </link>

  <!-- This is for the Kinect pointed straight on the third level. -->
  <!-- When the Kinect is  at -30 degrees, level, use xyz = "0.15 0 0.03" rpy= "0 0.5 0" instead. -->
  <joint name="kinect_joint" type="fixed">
    <origin xyz="0.15 0 0.1" 
            rpy="0 0 0" />
    <parent link="third_level_link" />
    <child link="kinect_link" />
  </joint>

  <link name="kinect_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <!-- <mesh filename="package://myBot/urdf/kinect.dae"/>   don't know why this does not work  -->
       <box size="0.06 .28 .08"/>
      </geometry> 

      <material name="yellow"/>
    </visual>
  </link>

  <joint name="kinect_camera_joint" type="fixed">
    <origin xyz="0.03 0 0.03" rpy="0 0 0" />
    <parent link="kinect_link" />
    <child link="kinect_camera_frame" />
  </joint>

  <link name="kinect_camera_frame">
  </link>

  <joint name="camera_depth_joint" type="fixed">
    <origin xyz="0 0 0" rpy="0 0 0" />
    <parent link="kinect_camera_frame" />
    <child link="camera_depth_frame" />
  </joint>

  <link name="camera_depth_frame">
  </link>

  <joint name="camera_depth_optical_joint" type="fixed">
    <origin xyz="0 0 0" rpy="-1.5707 0 -1.5707" />
    <parent link="camera_depth_frame" />
    <child link="camera_depth_optical_frame" />
  </joint>

  <link name="camera_depth_optical_frame">
  </link>

  <joint name="camera_rgb_joint" type="fixed">
    <origin xyz="0 -0.005 0" rpy="0 0 0" />
    <parent link="kinect_camera_frame" />
    <child link="camera_rgb_frame" />
  </joint>

  <link name="camera_rgb_frame">
  </link>

  <joint name="camera_rgb_optical_joint" type="fixed">
    <origin xyz="0 0 0" rpy="-1.5707 0 -1.5707" />
    <parent link="camera_rgb_frame" />
    <child link="camera_rgb_optical_frame" />
  </joint>

  <link name="camera_rgb_optical_frame">
  </link>

 <joint name="hokuyo_joint" type="fixed">
    <origin xyz="0.18 0 0.06" 
            rpy="0 0 0" />
    <parent link="second_level_link" />
    <child link="hokuyo_link" />
  </joint>

  <link name="hokuyo_link">
    <visual>
      <origin xyz="0 0 -0.02" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.035" length = "0.08"/>
      </geometry>
      <material name="yellow"/>
    </visual>
  </link>

 <joint name="Imu_joint" type="fixed">
    <origin xyz="0.11 0 0.02" 
            rpy="0 0 0" />
    <parent link="base_link" />
    <child link="Imu_link" />
  </joint>

  <link name="Imu_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size ="0.01 0.01 0.01"/>
      </geometry>
      <material name="blue">
      </material>
    </visual>
  </link>

 <joint name="stalk_joint" type="fixed">
    <origin xyz="-0.175 0.08 0.755" 
            rpy="0 0 0" />
    <parent link="second_level_link" />
    <child link="stalk_link" />
  </joint>

  <link name="stalk_link">
    <visual>
      <origin xyz="0 0 -0.3775"/>
      <geometry>
        <box size="0.025 0.025 0.755"/>
      </geometry>
      <material name="blue"/>
    </visual>
  </link>

 <joint name="nose_joint" type="fixed">
    <origin xyz="0.153 0 0" 
            rpy="0 0 0" />
    <parent link="fourth_level_link" />
    <child link="nose_link" />
  </joint>

  <link name="nose_link">
    <visual>
      <origin xyz="0 0 0.001" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.025" length = "0.002"/>
      </geometry>
      <material name="red"/>
    </visual>
  </link>


</robot>
